### PR TITLE
debounce AboutActivity. Fix #1917.

### LINKED ISSUE #1917: Options can be clicked twice in About Activity
#### Software and hardware versions 
Collect v1.12.2, 
Android v7.0, device used Moto M
Android v 4.4.2, device used Asus zenfone 5

#### Problem description
Options in `AboutAcivity` can be clicked twice. Fast double tapping of any of the option performs the action twice. 
  
#### Steps to reproduce the problem
1. Go to option menu.
2. Select `About` item.
3. Double click `Visit the ODK website` / `Join the ODK forum` / `Share ODK Collect` / `Leave a review` (quickly tap the option twice) 

On Moto M
1. `Visit the ODK website` 2 . `Join the ODK forum`
 

On Asus Zenfone 5

1. `Share ODK Collect` 2. `Leave a review`
 

#### Expected behavior
The options should be clickable only once.

#### Other information 
Related to `AboutActivity` in package `org.odk.collect.android.activities`.
The reproducibility of issue is device dependent

### PR DESCRIPTION
**Pull request by Thodoris1999**
Closes #1917 

#### What has been done to verify that this works as intended? 

Tested on device Meizu M2 (android 5.1)

#### Why is this the best possible solution? Were any other approaches considered?

At first I tried to set the flag back to false at the end of onClick, but this didn't work and it was still possible to create multiple intents. So, to make sure this doesn't happen I am setting the flag to false at onResume. I also considered making the debounce time-based, but this isn't compatible with web intents, as the time it takes for them to be executed varies a lot (up to two seconds on my device!)

#### Are there any risks to merging this code? If so, what are they?

If in the future someone wants to add an action that doesn't the foreground activity, clicking would be disabled until the activity is reloaded, if they forget to update the flag, so I added some comments as a reminder to avoid confusion.

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@Thodoris1999 
please check [Collect.allowClick()](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/application/Collect.java#L358) method. I think it should be used here.

File: collect_app/src/main/java/org/odk/collect/android/activities/AboutActivity.java [hunk 1]
```diff
@@ -83,52 +85,57 @@ private void initToolbar() {
...
-                        if (activity.name.contains("com.google.android")) {
-                            ComponentName name = new ComponentName(
-                                    activity.applicationInfo.packageName,
-                                    activity.name);
-                            intent.setComponent(name);
-                            startActivity(intent);
-                            intentStarted = true;
+        //debounce clicks
+        //actions that do not change the activity should set the flag to false once they finish!
+        if (!handlingClickFlag) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/AboutActivity.java]: please check [Collect.allowClick()](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/application/Collect.java#L358) method. I think it should be used here.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/activities/AboutActivity.java (hunk 1 above)
Comment (Thodoris1999) [reply collect_app/src/main/java/org/odk/collect/android/activities/AboutActivity.java]: I searched a bit for examples on how this method is used and I also noticed (for example in MainMenuActivity when the onClicks are set) that clicks are also call ActivityLogger.logAsAction(Object t, String context, String action)
Should I include this call on my PR? If so, what should I pass as the ”context” parameter?

**Comment by shobhitagarwal1612** [comment]
How about using Butterknife's DebounceOnClickListener?
https://github.com/JakeWharton/butterknife/blob/master/butterknife/src/main/java/butterknife/internal/DebouncingOnClickListener.java

**Review by grzesiek2010 — APPROVED** [review]

**Comment by grzesiek2010** [comment]
@opendatakit-bot label "needs testing"

**Comment by grzesiek2010** [comment]
> How about using Butterknife's DebounceOnClickListener?
https://github.com/JakeWharton/butterknife/blob/master/butterknife/src/main/java/butterknife/internal/DebouncingOnClickListener.java

Since we use that `allowClick()` method in other cases I think it should be used here too. If you find that solution from ButerKnife better/reliable/more legible you can refactor the code. I think it's important to use one solution in all cases so if something else works in all cases too but is better we can change it.

**Comment by shobhitagarwal1612** [comment]
Butterknife's DebounceOnClickListener does the same thing that is being done manually by `allowClick()`. But considering that for doing it with Butterknife, we have to bind the views using butterknife in all the places where it is being used. That will require a lot of code changes just to add debounce. So I do think for now `allowClick()` should be used.

**Comment by mmarciniak90** [comment]
Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0.

@lognaturel 
I was able to open two windows only on Android 4.4 and 5.1. 
I'm not sure that Android version or specific device is a condition.
I doubled `Share ODK Collect with your colleagues` window only if it was fresh app. 
After I opened it once, it works correctly the next time.

**Comment by Thodoris1999** [comment]
I can get 2 clicks on 5.1 too, but only when I am running it directly from android studio. It runs a lot slower that way so I am pretty certain that this is due to the 500ms window on Collect.allowClick() not being enough in this case. @mmarciniak90 Did you manage to get it in a case where the app responded quickly, but it still performed 2 clicks?

Suggestion: Instead of hardcoding the 500 ms window inside Collect.allowClick(), create a parameter and provide an increased window in this case. Refactoring the code is not too much of a problem.

**Comment by grzesiek2010** [comment]
@Thodoris1999 you can test 700ms for example. Not a big change but it might be enough.

**Comment by Thodoris1999** [comment]
After some more tests, I can't reproduce the problem at around 1 sec. I don't know if that makes the app unresponsive in other use cases of allowClick though....

Let me know what is best to do (create a parameter and provide 1 sec threshold only in this case or change to a larger amount in all cases) and I'll commit the changes after my class.

**Comment by mmarciniak90** [comment]
@Thodoris1999 
I did not run Collect form Android Studio. I builded apk by ./gradlew clean build.
I noticed this only on mentioned devices. 
Did you always install a new apk? I saw this problem only on fresh installation.

**Comment by Thodoris1999** [comment]
Then again, that may be just my Android studio being so slow, running the app normally barely causes any double clicks

@mmarciniak yes

**Comment by Thodoris1999** [comment]
I get no errors beyond 580ms I increased it a bit just to be safe. Can anyone test it on weaker devices?
@mmarciniak90 thanks for the advice! :D Can't believe how much time I have been wasting all this time...

**Comment by mmarciniak90** [comment]
@Thodoris1999 I was still able to reproduce problem  on both devices when set to 650 ms.
I checked also 700 ms with the same result.
Finally, following the advice from @grzesiek2010, I set 1000ms. 
Then I was not able to reproduce problem.

**Comment by mmarciniak90** [comment]
**Tested with success!**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Review by lognaturel — APPROVED** [review]

**Review by lognaturel — APPROVED** [review]

**Pull request merged**